File: plugins/modulation/chorus/chorus.js
// Chorus — variable delay (~18 ms base) modulated by a slow sine LFO,
// mixed with the dry signal. The 18 ms base keeps the wet path far
// enough from the input that the comb-filter notches never collapse
// into audible coloration — a tighter base would make this sound more
// like a flanger, which is what flanger.js is for.

function ChorusProcessor() {
  this.buf = new Array(20000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.bufLen = this.buf.length;
  this.writeIdx = 0;
  this.phase = 0;
}
ChorusProcessor.parameterDescriptors = [
  { name: 'rate',  label: 'Rate (Hz)',  defaultValue: 0.8,  minValue: 0.1, maxValue: 5 },
  { name: 'depth', label: 'Depth (ms)', defaultValue: 6,    minValue: 0,   maxValue: 15 },
  { name: 'mix',   label: 'Mix',        defaultValue: 0.45, minValue: 0,   maxValue: 1 }
];
ChorusProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var rate = parameters.rate[0];
  var depthMs = parameters.depth[0];
  var mix = parameters.mix[0];
  var baseSamp = sampleRate * 0.018;
  var depthSamp = sampleRate * depthMs * 0.001;
  var phaseInc = (2 * Math.PI * rate) / sampleRate;
  for (var i = 0; i < input.length; i++) {
    var lfo = Math.sin(this.phase);
    var read = this.writeIdx - (baseSamp + depthSamp * lfo);
    while (read < 0) read += this.bufLen;
    var i0 = Math.floor(read);
    var frac = read - i0;
    var i1 = (i0 + 1) % this.bufLen;
    var wet = this.buf[i0] * (1 - frac) + this.buf[i1] * frac;
    this.buf[this.writeIdx] = input[i];
    this.writeIdx = (this.writeIdx + 1) % this.bufLen;
    output[i] = input[i] * (1 - mix) + wet * mix;
    this.phase += phaseInc;
    if (this.phase > 2 * Math.PI) this.phase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('chorus', ChorusProcessor);

File: plugins/modulation/chorus/Chorus.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Chorus (native port) — LFO-modulated short delay (~18 ms base) blended
// with dry. Thickens a single voice into a small ensemble. 80s pad sheen
// at slow rates, watery shimmer at higher ones.
public final class Chorus implements VocalMonitorNativePlugin {
    private float[] buf;
    private int bufLen;
    private int writeIdx;
    private float phase;
    private int sampleRate = 44100;
    private float rate = 0.8f, depth = 6f, mix = 0.45f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        bufLen = 20000;
        buf = new float[bufLen];
        writeIdx = 0;
        phase = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "rate", "depth", "mix" }; }
    @Override public float parameterMin(String n) { return "rate".equals(n) ? 0.1f : 0f; }
    @Override public float parameterMax(String n) {
        if ("rate".equals(n)) return 5f;
        if ("depth".equals(n)) return 15f;
        return 1f;
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "rate":  return 0.8f;
            case "depth": return 6f;
            case "mix":   return 0.45f;
            default:      return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "rate":  return "Rate (Hz)";
            case "depth": return "Depth (ms)";
            case "mix":   return "Mix";
            default:      return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "rate":  rate = v; break;
            case "depth": depth = v; break;
            case "mix":   mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float baseSamp = sampleRate * 0.018f;
        final float depthSamp = sampleRate * depth * 0.001f;
        final float phaseInc = (float) (2.0 * Math.PI * rate / sampleRate);
        final float twoPi = (float) (2.0 * Math.PI);
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float[] b = buf;
        final int bL = bufLen;
        int w = writeIdx;
        float ph = phase;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float lfo = (float) Math.sin(ph);
            float read = w - (baseSamp + depthSamp * lfo);
            while (read < 0) read += bL;
            int i0 = (int) read;
            float frac = read - i0;
            int i1 = i0 + 1; if (i1 >= bL) i1 = 0;
            float wet = b[i0] * (1f - frac) + b[i1] * frac;
            b[w] = input[i];
            w++; if (w >= bL) w = 0;
            output[i] = input[i] * dry + wet * mixLocal;
            ph += phaseInc;
            if (ph > twoPi) ph -= twoPi;
        }
        writeIdx = w;
        phase = ph;
    }
}
